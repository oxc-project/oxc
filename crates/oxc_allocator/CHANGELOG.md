# Changelog

All notable changes to this crate will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does not adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) until v1.0.0.

## [0.6.0] - 2024-02-03

### Documentation

- Document behaviour of `Box`

## [0.5.0] - 2024-01-12

### Features

- Add `From` API (#1908)

## [0.4.0] - 2023-12-08

### Refactor

- Move to workspace lint table (#1444)

## [0.2.0] - 2023-09-14

### Refactor

- Improve code coverage a little bit
- Improve code coverage in various places (#721)

