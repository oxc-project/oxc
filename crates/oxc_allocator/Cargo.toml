[package]
name                   = "oxc_allocator"
version                = "0.11.0"
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
bumpalo = { workspace = true, features = ["collections"] }
serde   = { workspace = true }

[dev-dependencies]
serde_json = { workspace = true }
