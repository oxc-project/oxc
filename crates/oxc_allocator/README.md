# Oxc Allocator

A high-performance memory allocator using bump-based arena allocation for fast AST node creation.

## Overview

Oxc uses a bump-based memory arena for faster AST allocations. This crate provides an [`Allocator`] that manages memory efficiently by allocating objects in a single contiguous memory region, avoiding the overhead of individual heap allocations.

## Key Features

- **Bump allocation**: Fast allocation by simply incrementing a pointer
- **Arena-based memory management**: All allocations live for the same lifetime
- **Zero-copy data structures**: Efficient `Box`, `Vec`, `String`, and `HashMap` implementations
- **Optimal for AST operations**: Perfect for parse-transform-emit workflows

## Architecture

The allocator is designed specifically for AST processing workflows where:

1. A large number of nodes are allocated during parsing
2. All nodes have the same lifetime (tied to the AST)
3. Memory is released all at once when the AST is dropped

This approach is significantly faster than using the system allocator for AST operations.

## Features

- `serialize` - Enables serialization support for `Box` and `Vec` with `serde`
- `from_raw_parts` - Adds unsafe `from_raw_parts` method (not recommended for general use)
