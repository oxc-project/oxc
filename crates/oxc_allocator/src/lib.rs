use std::{
    convert::From,
    ops::{Deref, DerefMut},
};

mod arena;

pub use arena::{Box, String, Vec};
use bumpalo::Bump;

#[derive(Default)]
pub struct Allocator {
    bump: Bump,
}

impl From<Bump> for Allocator {
    fn from(bump: Bump) -> Self {
        Self { bump }
    }
}

impl Deref for Allocator {
    type Target = Bump;

    fn deref(&self) -> &Self::Target {
        &self.bump
    }
}

impl DerefMut for Allocator {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.bump
    }
}

#[cfg(test)]
mod test {
    use std::ops::Deref;

    use crate::Allocator;
    use bumpalo::Bump;

    #[test]
    fn test_api() {
        let bump = Bump::new();
        let allocator: Allocator = bump.into();
        #[allow(clippy::explicit_deref_methods)]
        {
            _ = allocator.deref();
        }
    }
}
