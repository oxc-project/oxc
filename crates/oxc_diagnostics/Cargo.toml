[package]
name                   = "oxc_diagnostics"
version                = "0.4.0"
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
thiserror = { workspace = true }
miette    = { workspace = true }

unicode-width = "0.1.11"
owo-colors    = { version = "3.5.0" }
textwrap      = { version = "=0.15.0" } # pinned to 0.15.0 to sync with miette
is-terminal   = { version = "0.4.10" }
