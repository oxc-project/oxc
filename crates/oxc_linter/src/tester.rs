use std::{
    env,
    path::{Path, PathBuf},
};

use oxc_allocator::Allocator;
use oxc_diagnostics::miette::NamedSource;
use oxc_diagnostics::{DiagnosticService, GraphicalReportHandler, GraphicalTheme};
use serde::Deserialize;
use serde_json::Value;

use crate::{
    rules::RULES, ESLintSettings, Fixer, LintOptions, LintService, LintServiceOptions, Linter,
    RuleEnum,
};

#[derive(Eq, PartialEq)]
enum TestResult {
    Passed,
    Failed,
    Fixed(String),
}

#[derive(Debug, Clone, Default)]
pub struct TestCase {
    source: String,
    config: Option<Value>,
    settings: Option<Value>,
    path: Option<PathBuf>,
}

impl From<&str> for TestCase {
    fn from(source: &str) -> Self {
        Self { source: source.to_string(), ..Self::default() }
    }
}

impl From<(&str, Option<Value>)> for TestCase {
    fn from((source, config): (&str, Option<Value>)) -> Self {
        Self { source: source.to_string(), config, ..Self::default() }
    }
}

impl From<(&str, Option<Value>, Option<Value>)> for TestCase {
    fn from((source, config, settings): (&str, Option<Value>, Option<Value>)) -> Self {
        Self { source: source.to_string(), config, settings, ..Self::default() }
    }
}

impl From<(&str, Option<Value>, Option<Value>, Option<PathBuf>)> for TestCase {
    fn from(
        (source, config, settings, path): (&str, Option<Value>, Option<Value>, Option<PathBuf>),
    ) -> Self {
        Self { source: source.to_string(), config, settings, path }
    }
}

pub struct Tester {
    rule_name: &'static str,
    rule_path: PathBuf,
    expect_pass: Vec<TestCase>,
    expect_fail: Vec<TestCase>,
    expect_fix: Vec<(String, String, Option<Value>)>,
    snapshot: String,
    current_working_directory: Box<Path>,
    import_plugin: bool,
    jest_plugin: bool,
    jsx_a11y_plugin: bool,
    nextjs_plugin: bool,
    react_perf_plugin: bool,
}

impl Tester {
    pub fn new<T: Into<TestCase>>(
        rule_name: &'static str,
        expect_pass: Vec<T>,
        expect_fail: Vec<T>,
    ) -> Self {
        let rule_path = PathBuf::from(rule_name.replace('-', "_")).with_extension("tsx");
        let expect_pass = expect_pass.into_iter().map(Into::into).collect::<Vec<_>>();
        let expect_fail = expect_fail.into_iter().map(Into::into).collect::<Vec<_>>();
        let current_working_directory =
            env::current_dir().unwrap().join("fixtures/import").into_boxed_path();
        Self {
            rule_name,
            rule_path,
            expect_pass,
            expect_fail,
            expect_fix: vec![],
            snapshot: String::new(),
            current_working_directory,
            import_plugin: false,
            jest_plugin: false,
            jsx_a11y_plugin: false,
            nextjs_plugin: false,
            react_perf_plugin: false,
        }
    }

    /// Change the path
    pub fn change_rule_path(mut self, path: &str) -> Self {
        self.rule_path = self.current_working_directory.join(path);
        self
    }

    pub fn with_import_plugin(mut self, yes: bool) -> Self {
        self.import_plugin = yes;
        self
    }

    pub fn with_jest_plugin(mut self, yes: bool) -> Self {
        self.jest_plugin = yes;
        self
    }

    pub fn with_jsx_a11y_plugin(mut self, yes: bool) -> Self {
        self.jsx_a11y_plugin = yes;
        self
    }

    pub fn with_nextjs_plugin(mut self, yes: bool) -> Self {
        self.nextjs_plugin = yes;
        self
    }

    pub fn with_react_perf_plugin(mut self, yes: bool) -> Self {
        self.react_perf_plugin = yes;
        self
    }

    pub fn expect_fix<S: Into<String>>(mut self, expect_fix: Vec<(S, S, Option<Value>)>) -> Self {
        self.expect_fix =
            expect_fix.into_iter().map(|(s1, s2, r)| (s1.into(), s2.into(), r)).collect::<Vec<_>>();
        self
    }

    pub fn test(&mut self) {
        self.test_pass();
        self.test_fail();
        self.test_fix();
    }

    pub fn test_and_snapshot(&mut self) {
        self.test();
        self.snapshot();
    }

    pub fn snapshot(&self) {
        let name = self.rule_name.replace('-', "_");
        insta::with_settings!({ prepend_module_to_snapshot => false, }, {
            insta::assert_snapshot!(name.clone(), self.snapshot, &name);
        });
    }

    fn test_pass(&mut self) {
        for TestCase { source, config, settings, path } in self.expect_pass.clone() {
            let result = self.run(&source, config, false, &settings, &path);
            let passed = result == TestResult::Passed;
            assert!(passed, "expect test to pass: {source} {}", self.snapshot);
        }
    }

    fn test_fail(&mut self) {
        for TestCase { source, config, settings, path } in self.expect_fail.clone() {
            let result = self.run(&source, config, false, &settings, &path);
            let failed = result == TestResult::Failed;
            assert!(failed, "expect test to fail: {source}");
        }
    }

    fn test_fix(&mut self) {
        for (test, expected, config) in self.expect_fix.clone() {
            let result = self.run(&test, config, true, &None, &None);
            if let TestResult::Fixed(fixed_str) = result {
                assert_eq!(expected, fixed_str);
            } else {
                unreachable!()
            }
        }
    }

    fn run(
        &mut self,
        source_text: &str,
        config: Option<Value>,
        is_fix: bool,
        settings: &Option<Value>,
        path: &Option<PathBuf>,
    ) -> TestResult {
        let allocator = Allocator::default();
        let rule = self.find_rule().read_json(config);
        let lint_settings: ESLintSettings = settings
            .as_ref()
            .map_or_else(ESLintSettings::default, |v| ESLintSettings::deserialize(v).unwrap());
        let options = LintOptions::default()
            .with_fix(is_fix)
            .with_import_plugin(self.import_plugin)
            .with_jest_plugin(self.jest_plugin)
            .with_jsx_a11y_plugin(self.jsx_a11y_plugin)
            .with_nextjs_plugin(self.nextjs_plugin)
            .with_react_perf_plugin(self.react_perf_plugin);
        let linter = Linter::from_options(options)
            .unwrap()
            .with_rules(vec![rule])
            .with_settings(lint_settings);
        let path_to_lint = if self.import_plugin {
            assert!(path.is_none(), "import plugin does not support path");
            self.current_working_directory.join(&self.rule_path)
        } else if let Some(path) = path {
            self.current_working_directory.join(path)
        } else {
            self.rule_path.clone()
        };

        let cwd = self.current_working_directory.clone();
        let paths = vec![path_to_lint.into_boxed_path()];
        let options = LintServiceOptions { cwd, paths, tsconfig: None };
        let lint_service = LintService::from_linter(linter, options);
        let diagnostic_service = DiagnosticService::default();
        let tx_error = diagnostic_service.sender();
        let result = lint_service.run_source(&allocator, source_text, false, tx_error);

        if result.is_empty() {
            return TestResult::Passed;
        }

        if is_fix {
            let fix_result = Fixer::new(source_text, result).fix();
            return TestResult::Fixed(fix_result.fixed_code.to_string());
        }

        let diagnostic_path = if self.import_plugin {
            self.rule_path.strip_prefix(&self.current_working_directory).unwrap()
        } else {
            &self.rule_path
        }
        .to_string_lossy();

        let handler = GraphicalReportHandler::new().with_theme(GraphicalTheme::unicode_nocolor());
        for diagnostic in result {
            let diagnostic = diagnostic.error.with_source_code(NamedSource::new(
                diagnostic_path.clone(),
                source_text.to_string(),
            ));
            handler.render_report(&mut self.snapshot, diagnostic.as_ref()).unwrap();
        }
        TestResult::Failed
    }

    fn find_rule(&self) -> &RuleEnum {
        RULES
            .iter()
            .find(|rule| rule.name() == self.rule_name)
            .unwrap_or_else(|| panic!("Rule not found: {}", &self.rule_name))
    }
}
