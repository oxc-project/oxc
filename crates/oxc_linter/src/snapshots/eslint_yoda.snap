---
source: crates/oxc_linter/src/tester.rs
snapshot_kind: text
---
  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (x <= 'foo' || 'bar' < x) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the left side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if ("red" == value) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the right side of ==.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (true === value) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the right side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (5 != value) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of !=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (5n != value) {}
   ·     ───────────
   ╰────
  help: Expected literal to be on the right side of !=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (null !== value) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the right side of !==.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if ("red" <= value) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (`red` <= value) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (`red` <= `${foo}`) {}
   ·     ─────────────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (`red` <= `${"red"}`) {}
   ·     ───────────────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (true >= value) {}
   ·     ─────────────
   ╰────
  help: Expected literal to be on the right side of >=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:12]
 1 │ var foo = (5 < value) ? true : false
   ·            ─────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:26]
 1 │ function foo() { return (null > value); }
   ·                          ────────────
   ╰────
  help: Expected literal to be on the right side of >.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (-1 < str.indexOf(substr)) {}
   ·     ────────────────────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (value == "red") {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the left side of ==.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (value == `red`) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the left side of ==.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (value === true) {}
   ·     ──────────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (value === 5n) {}
   ·     ────────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (`${"red"}` <= `red`) {}
   ·     ───────────────────
   ╰────
  help: Expected literal to be on the left side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:14]
 1 │ if (a < 0 && 0 <= b && b < 1) {}
   ·              ──────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a && a < 1 && b < 1) {}
   ·     ──────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (1 < a && a < 0) {}
   ·     ─────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ 0 < a && a < 1
   · ─────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:18]
 1 │ var a = b < 0 || 1 <= b;
   ·                  ──────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= x && x < -1) {}
   ·     ──────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:10]
 1 │ var a = (b < 0 && 0 <= b);
   ·          ─────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:10]
 1 │ var a = (b < `0` && `0` <= b);
   ·          ───────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (`green` < x.y && x.y < `blue`) {}
   ·     ─────────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[b] && a['b'] < 1) {}
   ·     ─────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[b] && a[`b`] < 1) {}
   ·     ─────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (`0` <= a[b] && a[`b`] < `1`) {}
   ·     ───────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[b] && a.b < 1) {}
   ·     ─────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[''] && a.b < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[''] && a[' '] < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[''] && a[null] < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[``] && a[null] < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[''] && a[b] < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[''] && a[b()] < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[``] && a[b()] < 1) {}
   ·     ──────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a[b()] && a[b()] < 1) {}
   ·     ───────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= a.null && a[/(?<zero>0)/] <= 1) {}
   ·     ───────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (3 == a) {}
   ·     ──────
   ╰────
  help: Expected literal to be on the right side of ==.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ foo(3 === a);
   ·     ───────
   ╰────
  help: Expected literal to be on the right side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ foo(a === 3);
   ·     ───────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ foo(a === `3`);
   ·     ─────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 <= x && x < 1) {}
   ·     ──────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:14]
 1 │ if ( /* a */ 0 /* b */ < /* c */ foo /* d */ ) {}
   ·              ───────────────────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:14]
 1 │ if ( /* a */ foo /* b */ > /* c */ 0 /* d */ ) {}
   ·              ───────────────────────
   ╰────
  help: Expected literal to be on the left side of >.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (foo()===1) {}
   ·     ─────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (foo()     === 1) {}
   ·     ───────────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:8]
 1 │ while (0 === (a));
   ·        ─────────
   ╰────
  help: Expected literal to be on the right side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:8]
 1 │ while (0 === (a = b));
   ·        ─────────────
   ╰────
  help: Expected literal to be on the right side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:8]
 1 │ while ((a) === 0);
   ·        ─────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:8]
 1 │ while ((a = b) === 0);
   ·        ─────────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (((((((((((foo)))))))))) === ((((((5)))))));
   ·     ─────────────────────────────────────────
   ╰────
  help: Expected literal to be on the left side of ===.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield(1) < a }
   ·                        ───────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield((1)) < a }
   ·                        ─────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:25]
 1 │ function *foo() { yield 1 < a }
   ·                         ─────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:28]
 1 │ function *foo() { yield/**/1 < a }
   ·                            ─────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield(1) < ++a }
   ·                        ─────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield(1) < (a) }
   ·                        ─────────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:3]
 1 │ x=1 < a
   ·   ─────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield++a < 1 }
   ·                        ───────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield(a) < 1 }
   ·                        ───────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:25]
 1 │ function *foo() { yield a < 1 }
   ·                         ─────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:28]
 1 │ function *foo() { yield/**/a < 1 }
   ·                            ─────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:24]
 1 │ function *foo() { yield++a < (1) }
   ·                        ─────────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:3]
 1 │ x=a < 1
   ·   ─────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ 0 < f()in obj
   · ───────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ 1 > x++instanceof foo
   · ───────
   ╰────
  help: Expected literal to be on the right side of >.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ x < ('foo')in bar
   · ───────────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ false <= ((x))in foo
   · ──────────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ x >= (1)instanceof foo
   · ────────
   ╰────
  help: Expected literal to be on the left side of >=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ false <= ((x)) in foo
   · ──────────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ x >= 1 instanceof foo
   · ──────
   ╰────
  help: Expected literal to be on the left side of >=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ x >= 1/**/instanceof foo
   · ──────
   ╰────
  help: Expected literal to be on the left side of >=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:2]
 1 │ (x >= 1)instanceof foo
   ·  ──────
   ╰────
  help: Expected literal to be on the left side of >=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ (x) >= (1)instanceof foo
   · ──────────
   ╰────
  help: Expected literal to be on the left side of >=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ 1 > x===foo
   · ─────
   ╰────
  help: Expected literal to be on the right side of >.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:1]
 1 │ 1 > x
   · ─────
   ╰────
  help: Expected literal to be on the right side of >.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:22]
 1 │ if (`green` < x.y && x.y < `blue`) {}
   ·                      ────────────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:16]
 1 │ if('a' <= x && x < 'b') {}
   ·                ───────
   ╰────
  help: Expected literal to be on the left side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if ('b' <= x && x < 'a') {}
   ·     ────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:4]
 1 │ if('a' <= x && x < 1) {}
   ·    ────────
   ╰────
  help: Expected literal to be on the right side of <=.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ if (0 < a && b < max) {}
   ·     ─────
   ╰────
  help: Expected literal to be on the right side of <.

  ⚠ eslint(yoda): Require or disallow "Yoda" conditions
   ╭─[yoda.tsx:1:5]
 1 │ {( t=='' )}
   ·    ─────
   ╰────
  help: Expected literal to be on the left side of ==.
